File: admin/src/api/order.ts
import { http, type ApiEnvelope } from '../utils/request'
import type { PaginatedData } from './admin-user'

export interface OrderUserSummary {
  id: number
  username: string
  email: string
  display_name: string | null
}

export interface AdminOrderItem {
  order_no: string
  user: OrderUserSummary
  status: 'pending' | 'cancelled' | 'closed'
  product_name: string
  plan_name: string
  billing_cycle: string
  total_amount_cents: number
  currency: string
  admin_note: string | null
  created_at: string
  cancelled_at: string | null
  closed_at: string | null
}

export interface AdminOrderDetail extends AdminOrderItem {
  user_note: string | null
  cancel_reason: string | null
  closed_reason: string | null
  product_no: string
  product_type: string
  product_summary: string | null
  plan_no: string
  plan_code: string
  plan_summary: string | null
  cpu_cores: number
  memory_mb: number
  system_disk_gb: number
  data_disk_gb: number
  bandwidth_mbps: number
  traffic_gb: number | null
  public_ip_count: number
  virtualization: string
  architecture: string
  price_cents: number
  original_price_cents: number | null
  quantity: number
  region_no: string
  region_code: string
  region_name: string
  template_no: string
  template_code: string
  template_name: string
  os_family: string
  os_distribution: string
  os_version: string
  os_architecture: string
}

export async function getOrders(params?: Record<string, unknown>) {
  const response = await http.get<ApiEnvelope<PaginatedData<AdminOrderItem>>>('/orders', { params })
  return response.data.data
}

export async function getOrderDetail(orderNo: string) {
  const response = await http.get<ApiEnvelope<AdminOrderDetail>>(`/orders/${orderNo}`)
  return response.data.data
}

export async function updateOrderAdminNote(orderNo: string, adminNote: string | null) {
  const response = await http.patch<ApiEnvelope<AdminOrderDetail>>(`/orders/${orderNo}/admin-note`, { admin_note: adminNote })
  return response.data.data
}

export async function cancelOrder(orderNo: string, reason?: string) {
  const response = await http.post<ApiEnvelope<AdminOrderDetail>>(`/orders/${orderNo}/cancel`, { reason })
  return response.data.data
}

export async function closeOrder(orderNo: string, reason?: string) {
  const response = await http.post<ApiEnvelope<AdminOrderDetail>>(`/orders/${orderNo}/close`, { reason })
  return response.data.data
}

File: server/internal/usecase/admin/dto/order.go
package dto

import "time"

type OrderListQuery struct {
	Page        int    `form:"page" validate:"omitempty,min=1"`
	PerPage     int    `form:"per_page" validate:"omitempty,min=1,max=100"`
	Status      string `form:"status" validate:"omitempty,oneof=pending provisioning fulfilled cancelled closed"`
	OrderNo     string `form:"order_no" validate:"omitempty,max=64"`
	UserKeyword string `form:"user_keyword" validate:"omitempty,max=128"`
	DateFrom    string `form:"date_from" validate:"omitempty,max=32"`
	DateTo      string `form:"date_to" validate:"omitempty,max=32"`
}

type OrderUserSummary struct {
	ID          uint64  `json:"id"`
	Username    string  `json:"username"`
	Email       string  `json:"email"`
	DisplayName *string `json:"display_name"`
}

type AdminOrderItem struct {
	OrderNo          string           `json:"order_no"`
	User             OrderUserSummary `json:"user"`
	Status           string           `json:"status"`
	ProductName      string           `json:"product_name"`
	PlanName         string           `json:"plan_name"`
	BillingCycle     string           `json:"billing_cycle"`
	NetworkTypeName  string           `json:"network_type_name"`
	TotalAmountCents uint64           `json:"total_amount_cents"`
	Currency         string           `json:"currency"`
	AdminNote        *string          `json:"admin_note"`
	CreatedAt        time.Time        `json:"created_at"`
	CancelledAt      *time.Time       `json:"cancelled_at"`
	ClosedAt         *time.Time       `json:"closed_at"`
}

type AdminOrderDetail struct {
	AdminOrderItem
	UserNote           *string `json:"user_note"`
	CancelReason       *string `json:"cancel_reason"`
	ClosedReason       *string `json:"closed_reason"`
	ProductNo          string  `json:"product_no"`
	ProductType        string  `json:"product_type"`
	ProductSummary     *string `json:"product_summary"`
	PlanNo             string  `json:"plan_no"`
	PlanCode           string  `json:"plan_code"`
	PlanSummary        *string `json:"plan_summary"`
	CPUCores           int     `json:"cpu_cores"`
	MemoryMB           int     `json:"memory_mb"`
	SystemDiskGB       int     `json:"system_disk_gb"`
	DataDiskGB         int     `json:"data_disk_gb"`
	BandwidthMbps      int     `json:"bandwidth_mbps"`
	TrafficGB          *int    `json:"traffic_gb"`
	PublicIPCount      int     `json:"public_ip_count"`
	Virtualization     string  `json:"virtualization"`
	Architecture       string  `json:"architecture"`
	PriceCents         uint64  `json:"price_cents"`
	OriginalPriceCents *uint64 `json:"original_price_cents"`
	Quantity           int     `json:"quantity"`
	RegionNo           string  `json:"region_no"`
	RegionCode         string  `json:"region_code"`
	RegionName         string  `json:"region_name"`
	NetworkTypeNo      string  `json:"network_type_no"`
	NetworkTypeCode    string  `json:"network_type_code"`
	NetworkTypeName    string  `json:"network_type_name"`
	TemplateNo         string  `json:"template_no"`
	TemplateCode       string  `json:"template_code"`
	TemplateName       string  `json:"template_name"`
	OSFamily           string  `json:"os_family"`
	OSDistribution     string  `json:"os_distribution"`
	OSVersion          string  `json:"os_version"`
	OSArchitecture     string  `json:"os_architecture"`
}

type OrderAdminNoteRequest struct {
	AdminNote *string `json:"admin_note" validate:"omitempty,max=1000"`
}

type OrderStatusRequest struct {
	Reason *string `json:"reason" validate:"omitempty,max=500"`
}
